File: balancer/service/navigate/src/core/Proc.cc
#include "Proc.h"

#include <boost/bind.hpp>
#include "define.h"
#include "tool/Util.h"

bool Proc::s_reload_flag = false;
bool Proc::s_stop_flag = false;

Proc::Proc(muduo::net::EventLoop& loop,
		   const std::string& config_file,
		   const unsigned int conf_id)
	:	_loop(loop), 
		_config_file(config_file),
		_conf_id(conf_id),
		_tcp_server(*this),
		_http_server(*this),
		_timer(*this)
{

}

Proc::~Proc()
{

}

int Proc::init()
{
	int ret = 0;
	std::string err;

	char config_json[Define::BUFFER_SIZE] = {0};
	ret = Util::file_2_bin(_config_file, config_json, sizeof(config_json)/sizeof(char));
	if(ret != 0)
	{
		B_LOG_ERROR << "open file config_file=" << _config_file;
		return -100;
	}

	err = _config.load(config_json);
	if(err.size() > 0)
	{
		B_LOG_ERROR << "config load err=" << err;
		return -200;
	}

	B_LOG_INFO << _config.to_string();
	return 0;
}

void Proc::start()
{
	B_LOG_INFO;
	_tcp_server.start();
	_http_server.start();
	_loop.runEvery(0.1, boost::bind(&Timer::check_timeout, &_timer));
}

void Proc::quit()
{
	B_LOG_INFO;
	_loop.quit();
}

void Proc::loop()
{
	B_LOG_INFO;
	_loop.loop();
}

void Proc::logging()
{
	_log.init(_config.log.level, _config.log.console, _config.log.file_path, _config.log.roll_size,
			  _config.log.thread_safe, _config.log.flush_interval, _config.log.check_every_n);
}

void Proc::check_flag()
{
	if(s_reload_flag)
	{
		s_reload_flag = false;
		B_LOG_ERROR << "reload start";

		if(reload())
		{
			B_LOG_ERROR << "reload end";
			logging();
		}
		else
		{
			B_LOG_ERROR << "reload failed";
		}	
	}

	if(s_stop_flag)
	{
		s_stop_flag = false;
		B_LOG_ERROR << "stop start";
		quit();
	}
}

bool Proc::reload()
{
	// 有错误，则不重新加载
	int ret = 0;
	std::string err;

	char config_json[Define::BUFFER_SIZE] = {0};
	ret = Util::file_2_bin(_config_file, config_json, sizeof(config_json)/sizeof(char));
	if(ret != 0)
	{
		B_LOG_ERROR << "open file config_file=" << _config_file;
		return false;
	}

	err = _config.reload(config_json);
	if(err.size() > 0)
	{
		B_LOG_ERROR << "config load err=" << err;
		return false;
	}

	B_LOG_INFO << _config.to_string();
	return true;
}

File: balancer/service/proxy/src/core/proc.py
# -*- coding: utf-8 -*-

import time
import asyncio
from concurrent import futures

from define import Define
from log import log
from config import config
from config import service_config
from core import seq
from core import timer
from core import owner
from core import inservice_service
from core import tcp_server

g_proc = None


class Proc:
    s_reload_flag = False
    s_stop_flag = False

    @staticmethod
    def set_reload(b):
        Proc.s_reload_flag = b

    @staticmethod
    def set_stop(b):
        Proc.s_stop_flag = b

    def __init__(self, config_file):
        global g_proc
        g_proc = self

        self.conn_map = {}
        self.loop = None
        self.pool = None

        self.config_file = config_file
        self.config = config.Config()
        self.sc = service_config.ServiceConfig()
        self.seq = seq.Seq()
        self.owner = owner.Owner()
        self.inservice_service = inservice_service.InserviceService()

    def init(self):
        with open(self.config_file) as fd:
            config_json = fd.read(Define.BUFFER_SIZE)
        b = self.config.load(config_json)
        if not b:
            print('config load err')
            return -100

        print(self.config)
        return 0

    def start(self):
        log.g_logger.info('start')
        if self.pool is None:
            self.pool = futures.ThreadPoolExecutor(self.config.proxy.thread_size)

        if self.loop is None:
            self.loop = asyncio.get_event_loop()
            self.loop.call_later(0, timer.Timer.check_timeout, self)

    def quit(self):
        log.g_logger.info('quit')
        self.loop.stop()

    def looping(self):
        log.g_logger.info('looping')
        coro = asyncio.start_server(tcp_server.on_connection,
                                    self.config.net.tcp.ip,
                                    self.config.net.tcp.port,
                                    loop=self.loop)
        server = self.loop.run_until_complete(coro)
        log.g_logger.info('Serving on {}'.format(server.sockets[0].getsockname()))

        # Server requests until Ctrl+C is pressed
        try:
            self.loop.run_forever()
        except KeyboardInterrupt:
            pass

        # Close the server
        server.close()
        self.loop.run_until_complete(server.wait_closed())
        self.loop.close()

    def logging(self):
        log.logger(self.config)

    def check_flag(self):
        if Proc.s_reload_flag:
            Proc.s_reload_flag = False
            log.g_logger.error('reload start')
            if self.reload():
                log.g_logger.error('reload end')
            else:
                log.g_logger.error('reload failed')

        if Proc.s_stop_flag:
            Proc.s_stop_flag = False
            log.g_logger.error('stop start')
            self.quit()

    def reload(self):
        with open(self.config_file) as fd:
            config_json = fd.read(Define.BUFFER_SIZE)
        b = self.config.reload(config_json)
        if not b:
            log.g_logger.error('config load err')
            return False

        log.g_logger.info(self.config)
        return True

    def on_check_idle(self):
        now = time.time()
        timeout_list = []
        for k, v in self.conn_map.items():
            if now - v.update_time >= self.config.proc.tcp_server_idle:
                if v.close():
                    timeout_list.append(k)

        for l in timeout_list:
            self.conn_map.pop(l)
